🎨 레스토랑 상세 페이지 구현하기 ( 진행중... )

knoa0405 opened the pull request:
강의 내용까지만 해서 PR 보냅니다!
차차 과제에 맞게 바꿔가면서 PR 계속 보내겠습니다~!

knoa0405 committed:
setting webpack for index / Add RestaurantsPage

knoa0405 committed:
[Red]Add test code RestaurantsPage

knoa0405 committed:
[Green] Implement RestaurantsPage

knoa0405 committed:
Add route in App

knoa0405 committed:
🔥 Remove useless App test code

knoa0405 committed:
Add router and  Home/about/NotFound components

knoa0405 committed:
[Red] Add HomePage test code

knoa0405 committed:
[Green] Implement HomePage

knoa0405 committed:
[Green] Replace HomePage in App

knoa0405 committed:
[Red] Add test code about page

knoa0405 committed:
[Green] Implement About page

knoa0405 committed:
[Red] Add not found page

knoa0405 committed:
[Green] implement Not found Page

knoa0405 committed:
[Refactor] code App apply react-router-dom

knoa0405 committed:
[Refactor] code home page apply Link

knoa0405 committed:
[Green] Add test code in Homepage MemoryRouter

knoa0405 committed:
[Refactor] Move browserRouter from App to index

knoa0405 committed:
[Green] Add test code MemoryRouter in App

knoa0405 committed:
[Red] Add test code in App related about Page

knoa0405 committed:
[Green] Implement About page

knoa0405 committed:
[Refactor] duplicated test code in App

knoa0405 committed:
Add test code and header in App

knoa0405 committed:
[Refactor] Add Link in header

knoa0405 committed:
[Red] Add test code about Header

knoa0405 committed:
[Green] Implement Header

knoa0405 committed:
[Red] test code in AboutPage

knoa0405 committed:
[Green] Implement AboutPage

knoa0405 committed:
Add Link in Restaurants

knoa0405 committed:
[Red] Add test code about RestaurantsDetailsPage

knoa0405 committed:
[Green] Implement RestaurantDetailsPage

knoa0405 committed:
[Refactor] code in App

knoa0405 committed:
[Red] Add test code in RestaurantsContainer

knoa0405 committed:
[Red] fix test code in RestaurantsContainer

knoa0405 committed:
[Refactor] code in RestaurantsContainer

knoa0405 committed:
[Red] Add actions test code

knoa0405 committed:
Add setRestaurantDetails for pass action test

knoa0405 committed:
Add api for pass fetchRestaurantsDetails test

knoa0405 committed:
[Red] Add test setRestaurantDetails in reducer

knoa0405 committed:
[Green] Implement reducer / setRestaurantDetails

knoa0405 committed:
[Green] Implement api mock for pass test code

fetchRestaurantDetails not function error test 통과 시키기 위해
 __mock__ 파일에 api.js 에 추가

knoa0405 committed:
[Green] Implement code about RestaurantDetailsPage

To pass the test, I added the restaurantsDetails
 variable to the reducer function and the test function
 and implement the restaurantsDetailsPage.

knoa0405 committed:
[Red] Add test code RestaurantDetailsPageContainer

knoa0405 committed:
[Green] Implement RestaurantDetailsPageContainer

knoa0405 committed:
[Refactor] test code about RestaurantDetailsPage

knoa0405 committed:
Add  test code in App

Knoa0405 commented:
### url 파라미터가 일치하지 않아 test 통과가 안되는 문제가 있습니다.

1. 다음 PR까지 /details 에서 /restaurants/1 으로 가도록 수정해보겠습니다.

2. 리팩토링과 테스트 케이스들도 계속 추가해보겠습니다.

- 1번과 관련해서 방향을 알려주시면 반영해보겠습니다~!
   match.params 로 가능한 것 같은데 맞는 방향인가요?

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,78 @@
-  const dispatch = jest.fn();
+describe('App', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
```
beforeEach에서 모든 mocking함수들을 초기화하고 있어서 안써줘도 될 것 같아요. 만약 dispatch함수가 공통으로 다른 곳에서 쓰이기 위해 여기 바깥에서 선언된다면 필요할거에요

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,78 @@
+      regions: [{ id: 1, name: '서울' }],
+      categories: [],
+      restaurants: [],
+      restaurantDetails: {
```
Detail은 하나니까 단수형으로 써주는건 어떨까요?

hannut91 commented on src/RestaurantDetailsPage.jsx:
```diff
@@ -0,0 +1,23 @@
+      <ul>
+        {menuItems.map((menuItem) => (
+          <li key={menuItem.id}>{menuItem.name}</li>
+        ))}
```
구조 분해 할당으로 반복적으로 사용되는 menuItem을 제거해주는건 어떨까요?

hannut91 commented on src/RestaurantDetailsPage.test.jsx:
```diff
@@ -0,0 +1,35 @@
+  });
+
+  context('without restaurant details', () => {
+    it("renders 'Loading...'", () => {
```
너무 구체적인 구현을 테스트 이름에 포함되어 있는 것 같아요. 구현이 조금만 바껴도 테스트 이름까지 영향을 쉽게 미칠것 같아여

hannut91 commented on src/RestaurantDetailsPageContainer.jsx:
```diff
@@ -0,0 +1,21 @@
+      name={name}
+      address={address}
+      menuItems={menuItems}
+    />
```
이름, 주소, 메뉴들을 우리가 레스토랑이라고 부르니까 props를 레스토랑이라고 전달하는건 어떨까요?

hannut91 commented on src/RestaurantsPage.jsx:
```diff
@@ -0,0 +1,27 @@
+
+  useEffect(() => {
+    dispatch(loadInitialData());
+  });
```
어떤 값이 변할 때 useEffect훅을 실행할 지 명시해주는 게 좋을 것 같아요. 만약 컴포넌트가 마운트 될 때 한 번만 실행하고 싶다면 빈 배열을 전달해주는게 좋을 것 같습니다

hannut91 commented on src/reducer.js:
```diff
@@ -45,6 +50,19 @@ const reducers = {
+
+    return {
+      ...state,
+      restaurantDetails: {
```
restaurantDetails를 바로 할당해도 괜찮을까요??

hannut91 commented:
말씀하신대로 match.params도 가능하고 훅을 사용한다면 `useParams`를 사용하는데, mocking하기가 어려워서 간단하게 구현할 때는 

```js
```

knoa0405 committed:
[Refactor] code for pass test

1.첫번째 피드백 반영
- 네이밍, 비구조화 할당 변경
2. match.params를 이용해서 url 변경
3. useEffect 한번 호출을 위해 [] 추가
4. 웹팩 변경 main.js 404 ( 에러 해결 )

Knoa0405 commented:
### 리팩토링에 집중하다보니 단위별로 세세하게 커밋을 나누는걸 깜빡했습니다..

1.첫번째 피드백 반영
- 네이밍, 비구조화 할당 등 변경

2. match.params를 이용해서 url 변경 ( /details => /restaurants/:id )

match.params의 프로퍼티 id로 받아와서 그걸로 loadRestaurantDetail 
action 에 넣어서 fetch에 사용

( 기존에는 RestaurantContainer 안에서 해당 li에 onClick 이벤트를 부여하여
handleClick 함수 안에서 dispatch하여 id를 받아와 load, fetch 했었습니다. )

3. useEffect 한번 호출을 위해 [] 추가

4. 웹팩 변경html main.js 404 ( 에러 해결 )
- 스택오버플로우 참고했습니다. (https://stackoverflow.com/questions/51051490/react-webpack-dev-server-main-js-not-found)

( 중복제거, 네이밍, 테스트 추가 등 ) 계속보면서 리팩토링 해보겠습니다~!

hannut91 commented:
잘 해결해 주신 것 같아요 ㅎㅎ `html`에 `<base />`태그를 추가해서 문제를 해결할 수도 있습니다. webpack에 `publicPath`를 설정해서 문제를 해결할 수도 있습니다.

See also
  - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

hannut91 commented on src/services/api.test.js:
```diff
@@ -49,4 +51,18 @@ describe('api', () => {
+
+  describe('fetchRestaurantDetails', () => {
+    beforeEach(() => {
+      mockFetch({ restaurantDetail: { name: '마법사주방', address: '부산시 해운대구', menuItems: [] } });
```
`restaurantDetail`의 모양이 어떤지 보여줄려고 하신 것 같아요. 그래도 여기에 나열하는 것보다는 위에 선언해서 보여주는게 좋을 것 같습니다

hannut91 commented on src/reducer.test.js:
```diff
@@ -108,4 +114,20 @@ describe('reducer', () => {
     });
   });
+
+  describe('setRestaurantsDetails', () => {
+    it('changes selected category', () => {
+      const initialState = {
+        restaurantDetail: {},
+      };
+
+      const restaurantDetail = {
+        name: '양천 주가', address: '부산시 해운대구', menuItems: [],
+      };
+
+      const state = reducer(initialState, setRestaurantDetails(restaurantDetail));
+
+      expect(state.restaurantDetail).toEqual(restaurantDetail);
+    });
+  });
```
테스트에서는 카테고리를 변경한다고 했는데, 레스토랑 상세가 바뀌고 있는 것 같아요.

hannut91 commented on src/actions.test.js:
```diff
@@ -83,4 +86,22 @@ describe('actions', () => {
     });
   });
+
+  context('with restaurant details', () => {
```
이건 특정 상황이 다를 때 테스트하는 것이 아니라 `loadRestaurantDetails`를 테스트하는 것 같아요. 
어떤 테스트를 하고 있는건지 써주는게 좋을 것 같아요

hannut91 commented on src/RestaurantDetailsPageContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+      const { container } = render((
+        <MemoryRouter>
+          <RestaurantDetailsPage
+            restaurant={restaurantDetail}
```
props를 restaurant로 하기로 했으니 변수이름도 같이 `restaurant`로 변경하는건 어떨까요?

hannut91 commented on src/RestaurantDetailsPage.test.jsx:
```diff
@@ -0,0 +1,33 @@
+    </MemoryRouter>
+  ));
+
+  context('with restaurant detail', () => {
+    it('renders restaurant details page', () => {
+      const { container } = renderDetailsPage({ name: '양천 주가', address: '서울 강남구', menuItems: [{ id: 1, name: '밥' }] });
+
+      expect(container).toHaveTextContent('양천 주가');
+    });
+  });
+
+  context('without restaurant detail', () => {
+    it('shows loading text', () => {
+      const { container } = renderDetailsPage({ name: '', address: '', menuItems: [] });
+
+      expect(container).toHaveTextContent('Loading...');
+    });
+  });
```
```js
context('with restaurant detail', () => {
  const restaurant = {
    name: '양천 주가',
    address: '서울 강남구',
    menuItems: [{ id: 1, name: '밥' }],
  };

  it('renders restaurant details page', () => {
    const { container } = renderDetailsPage(restaurant);

    expect(container).toHaveTextContent('양천 주가');
  });
});

context('without restaurant detail', () => {
  const restaurant = {
    name: '',
    address: '',
    menuItems: [],
  };

  it('shows loading text', () => {
    const { container } = renderDetailsPage(restaurant);

    expect(container).toHaveTextContent('Loading...');
  });
});
```

이렇게하면 각 Given이 어떻게 다른지 더 명확하게 보일 것 같아요

hannut91 commented on src/RestaurantDetailsPage.test.jsx:
```diff
@@ -0,0 +1,33 @@
+    });
+  });
+
+  context('without restaurant detail', () => {
```
레스토랑 상세가 없다기 보다 비어있다는 표현이 더 좋을 것 같아요

hannut91 commented on src/RestaurantDetailsPage.jsx:
```diff
@@ -0,0 +1,23 @@
+      </p>
+      <h2>메뉴</h2>
+      <ul>
+        {menuItems.map(({ name: itemName, id }) => (
```
위에서 이미 `name`을 사용하고 있어서 `itemName`으로 변경하셨군요. 위의 `name`이 변수의 스코프가 더 길기 때문의 위의 이름을 더 구체적인 `restaurant.name` 같은 이름으로 변경하고 여기를 `name`으로 사용하는게 좋을 것 같아요. 

혹은 `menuItems`라는 컴포넌트로 분리하면 이런 문제가 발생하지 않겠죠?

hannut91 commented on src/HomePage.test.jsx:
```diff
@@ -0,0 +1,17 @@
+
+import HomePage from './HomePage';
+
+describe('HomePage', () => {
+  it('renders home page', () => {
+    render((
+      <MemoryRouter>
+        <HomePage />
+      </MemoryRouter>
+    ));
+  });
+});
```
간단한 테스트로 시작해서 좋네요 ㅎㅎ 또 다른 일을 하는 것이 있나요?

knoa0405 committed:
[Refactor] test code fetch restaurantDetail

knoa0405 committed:
[Refactor] test code in reducer

Modify wrong test comment

knoa0405 committed:
[Refactor] test code in actions

knoa0405 committed:
[Refactor] changes name in RestaurantDetailsPageCo

knoa0405 committed:
[Refactor] test code in RestaurantDetails

RestaurantDetailPage
context 에서 Given 을 명확히!
Given - When - Then

knoa0405 committed:
[Refactor] test code text in RestaurantDetailPage

knoa0405 committed:
[Refactor] Separate menuItems from Detail Page

knoa0405 committed:
[Refactor] Add test code in menuItems

Knoa0405 commented:
## 시간을 못내서 작고 늦게나마 피드백 해주신 것들만 반영해서 PR 보냈습니다!

내일은 제대로 시간 잡고 리팩토링 해보겠습니다!

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,76 @@
+  });
 
-  expect(dispatch).toBeCalled();
+  context('with invalid path', () => {
```
저는 `invalid`하다는 것은 값이 정말 유효하지 않을 때 사용하는 말 같아요. 예를들면 url이 `http://example.com\xxx`같은 것 처럼요

hannut91 commented on src/RestaurantMenuItems.jsx:
```diff
@@ -0,0 +1,14 @@
+        {menuItems.map(({ name: itemName, id }) => (
+          <li key={id}>{itemName}</li>
+        ))}
+      </ul>
```
이제 name을 `itemName`으로 바꿀 필요 없이 `name`을 사용해도 문제가 발생하지 않을거에요. 그리고 변수를 사용하는 순서가 `id`, `name`순이니 구조 분해 할당을 할 때 순서대로 써주면 더 이해하기 쉬울거에요

knoa0405 committed:
[Refactor] code in App, RestaurantMenuItems

knoa0405 committed:
Add test code in HomePage / NotFoundPage

hannut91 commented on src/AboutPage.test.jsx:
```diff
@@ -0,0 +1,15 @@
+  it('renders about page', () => {
+    const { getByText } = render(<AboutPage />);
+
+    expect(getByText('About')).not.toBeNull();
```
이 테스트는 아래의 테스트와 중복되는 것 같아요

hannut91 commented on src/AboutPage.test.jsx:
```diff
@@ -0,0 +1,15 @@
+
+    expect(getByText('About')).not.toBeNull();
+
+    expect(getByText('About 페이지입니다.')).not.toBeNull();
```
단순히 텍스트가 있는지 확인할 때는 다음과 같이 하는 것도 좋을 것 같아요

```js
const { container } = render(<AboutPage />);

expect(container).toHaveTextContent('About 페이지입니다.');
```

hannut91 reviewed (approved):
고생하셨습니다 ㅎㅎ `html-webpack-plugin`까지 활용하며 과제를 완성해주셨네요.

이번 과제를 통해 무엇을 배우셨나요?

knoa0405 committed:
[Refactor] test code in AboutPage

hannut91 commented:
채널에 회고를 남겨주셔서 머지합니다 ㅎㅎ 고생하셨습니다
